File: Longest Palindromic Substring/Longest_Palindromic_Substring.py
# 1004ms 77.20%
# 基本思路是，对于每个字符和两个字符的中间，都可能成为一个回文子字符串的中心
# 因此，使用线性的扫描，对每个中心进行查找
class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s:
            return ''
        len_str, max_length, index = len(s), 0, 0.5
        while index < len_str - 1:
            left = int(index - 0.5)
            right = int(index) + 1
            length = 1 if int(index) == index else 0
            
            while left >= 0 and right <= len_str - 1:
                if s[left] == s[right]:
                    length += 2
                    if length > max_length:
                        max_length = length
                        max_sub_string = s[left:right + 1]
                    left -= 1;right += 1
                else:
                    break
            index += 0.5
        return s[0] if max_length is 0 else max_sub_string

File: Longest Palindromic Substring/Longest_Palindromic_Substring.cpp
// Runtime: 48 ms, faster than 52.20% of C++ online submissions for Longest Palindromic Substring.
// Memory Usage: 16 MB, less than 39.08% of C++ online submissions for Longest Palindromic Substring.

class Solution {
public:
	string longestPalindrome(string s) {
		// 边界条件处理
		string res = "";
		if (s.length() == 0) return res;

		int maxLength = 0;
		for (int center = 0; center < s.length(); ++center) {
			int curLength[] = { 1, 0 }, leftPtr[] = { center, center + 1 }, rightPtr[] = { center, center };
			for (int i = 0; i < 2; ++i) {
				while (--leftPtr[i] >= 0 && ++rightPtr[i] < s.length() && s[leftPtr[i]] == s[rightPtr[i]])
					curLength[i] += 2;

				if (curLength[i] > maxLength) {
					res = s.substr(++leftPtr[i], curLength[i]);
					maxLength = curLength[i];
				}
			}
		}
		return res;
	}
};